### DISCUSSION TITLE
Has anyone managed to flash an AT32F415 Gotek in Linux?

### DISCUSSION THREAD

**Discussion by Lee-Adamson** [keirf/flashfloppy#698] [General]
Gentlemen.  I am about at my wit's end.

SFRKC30.AT2, with the AT32F415.  Devuan Chimaera (basically Debian 11.1 without systemd).  AMD64.  Windows isn't an option.  I'll flip these units on ebay and buy some with flashfloppy already installed before I install Windows on something just to flash a gotek.

Trying to flash over USB.  Downloaded the "ISP Console for Linux".  Zip file.  Ok, cool I guess.  Extract.

Head->desk for a while in library hell.  Two beers later, it turns out that it ships with an obsolete version of QT that is dynamically linked against an obsolete release of libicu.  I eventually delete their QT and symlink in my system QT and the software starts working.

Disable flash access protection.  Check.
Disable erase and program protection.  Check.
Try to flash+verify flashfloppy.
Segmentation fault (core dumped).
One more beer won't hurt.

```
root@Jormungandr:/home/leaf/ISP_Console_V3.0.03_Linux/Artery_ISP_Console_Linux_V3.0.03# ./AT32_ISP_Console -dfu -d --fn ../flashfloppy-3.36/flashfloppy-3.36/hex/flashfloppy-at415-st105-3.36.hex
------------------------------------------------------------
Artery ISP Programmer V3.0.02
Support USART and DFU
------------------------------------------------------------

DFU Connect......
Device:  AT32F415KBU7-4    Flash:  128 KB
PID:  700301C6      BID:  4601    Version:  3.2
Device connected Successfully

Segmentation fault (core dumped)
root@Jormungandr:/home/leaf/ISP_Console_V3.0.03_Linux/Artery_ISP_Console_Linux_V3.0.03# gdb ./AT32_ISP_Console core
GNU gdb (Debian 10.1-1.7) 10.1.90.20210103-git
Copyright (C) 2021 Free Software Foundation, Inc.                                                                                                            
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./AT32_ISP_Console...
[New LWP 896]
[New LWP 897]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `./AT32_ISP_Console -dfu -d --fn ../flashfloppy-3.36/flashfloppy-3.36/hex/flashf'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:413
413     ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S: No such file or directory.
[Current thread is 1 (Thread 0x7f75262325c0 (LWP 896))]
(gdb) bt
#0  __memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:413
#1  0x00007f752922db4c in memcpy (__len=536, __src=<optimized out>, __dest=<optimized out>) at /usr/include/x86_64-linux-gnu/bits/string3.h:53
#2  CImage::FilterImageForOperation (this=0x2054920, pMapping=<optimized out>, Operation=Operation@entry=2, bTruncateLeadFF=<optimized out>)
    at Image.cpp:1292
#3  0x00007f75292280c1 in FILES_FilterImageForOperation (Handle=<optimized out>, pMapping=<optimized out>, Operation=Operation@entry=2, 
    bTruncateLeadFFForUpgrade=<optimized out>) at Files.cpp:89
#4  0x00000000004239a2 in CProgramFile::Open (this=this@entry=0x7ffc149dc8d0, 
    strPath="../flashfloppy-3.36/flashfloppy-3.36/hex/flashfloppy-at415-st105-3.36.hex", obDevice=obDevice@entry=0x1c184f0) at CFiles.cpp:237
#5  0x000000000041c31e in DownloadFileAndAddressCheck (fileName="../flashfloppy-3.36/flashfloppy-3.36/hex/flashfloppy-at415-st105-3.36.hex", 
    downlaodAddress=downlaodAddress@entry=134217728, m_obBootLoader=m_obBootLoader@entry=0x1c182c0) at main.cpp:1789
#6  0x000000000041c878 in Download (argc=4, argv=argv@entry=0x7ffc149dcd10, arg_index=@0x7ffc149dcb3c: 3, erasedFlag=erasedFlag@entry=false, 
    m_obCCom=0x1c187c0) at main.cpp:778
#7  0x0000000000407cae in main (argc=4, argv=0x7ffc149dcd10) at main.cpp:2108
(gdb) q
```

Looks like they didn't strip the symbol table.  Too bad there's no source code.

Tempted to try it over serial.  Problem is, I don't have any TTL-level USB-to-serial dongles.  When programming ESP-01 units and the like, I've always just used a raspberry pi, since it already has a 3.3v serial port on the GPIO header anyway.  But the only scraps they throw us are a precompiled amd64 binary linked against some obsolete libraries, so flashing from a Pi isn't gonna happen this time.

I do have USB-to-RS232-level dongles, though, along with MAX3232 ICs, a bench supply, and a breadboard.

But I figured I ought to ask in here before I do something that ends up requiring more beer:

Have any of y'all managed to flash a '415 unit from Linux?  I'd appreciate any clue-sticks you gentlemen could hit me with.

Many thanks.

**Comment by Lee-Adamson** [comment]
I figured it out, using dfu-util. Sorry for wasting yourall's time!  I should have dug deeper before giving up.

**Reply by Lee-Adamson** [reply]
For reference, if anyone comes looking for the same info.....

What worked for me to flash from Linux:

1. Stay as far away from the official ISP Console Programmer software as possible.  This is the most important step.
2. Become root.
3. apt-get install dfu-util
4. download the flashfloppy release and unpack it.  There will be a dfu/ directory containing the firmware file.  You want to use the .dfu file, not the .hex file.
5. set the programming jumper on the gotek as per wiki.
6. attach gotek to linux box via usb-a to usb-a cable.
7. dfu-util --list # This should display two entries for one device (just different alt numbers). If more than one device is shown, you'll have to add the --device option as per the dfu-util manpage.
8. dfu-util -D /path/to/flashfloppy-at415-st105-3.36.dfu -a 0 -s 0x08000000:unprotect:force # this will fail with an error, but will unprotect the internal flash in the process.
9. dfu-util -D /path/to/flashfloppy-at415-st105-3.36.dfu -a 0 # Now that the flash is unlocked, the device will be programmed.
10. unplug usb cable.  Move programming jumper back to non-programming position.  Apply 5v to appropriate berg header pins.  7-segment display should read "F-F" if the programming was successful.

I'm not sure what is up with step 8, but it works to unprotect the flash so good enough I reckon.

Again, I apologize for my dumbness, and for asking for help prematurely.  I hope that this info helps someone else out.

**Reply by keirf** [reply]
Thanks for this step-by-step guide. I have added a link to it from the wiki.

**Reply by Lee-Adamson** [reply]
There was an error in the commandline in step 8, but it has been fixed.  Too many zeroes.  Sorry about that!

**Reply by GrantWagner** [reply]
Thank you for this, I just now attempted this process on a brand new gotek I purchased on amazon. Worked like a charm and got me running.

**Reply by wkjagt** [reply]
I just tried this on my Mac and it works flawlessly. `dfu-util` can be installed through `brew` (instead of `apt-get`) on Mac. The rest of this guide worked without any changes 🙌

**Reply by theflynn49** [reply]
I'm late to the party, but thank you anyway !

**Reply by pauljsymonds** [reply]
`Match vendor ID from file: 0483
Match product ID from file: df11
Multiple alternate interfaces for DfuSe file
Opening DFU capable USB device...
Device ID 0483:df11
Device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Interface #0 ...
Determining device status...
DFU state(10) = dfuERROR, status(15) = Device stalled an unexpected request
Clearing status
Determining device status...
DFU state(2) = dfuIDLE, status(0) = No error condition is present
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
File contains 1 DFU images
Parsing DFU image 1
Target name: ST...
Image for alternate setting 0, (2 elements, total size = 116644)
Setting Alternate Interface #0 ...
Parsing element 1, address = 0x08000000, size = 30316
Erase   	[                         ]   0%            0 bytesdfu-util: dfuse_download: libusb_control_transfer returned -7 (LIBUSB_ERROR_TIMEOUT)
dfu-util: Error during special command "ERASE_PAGE" download

Gotek-Setup/flashfloppy-3.44/dfu took 5s 
❯ 
`
Why would this happen?

**Reply by keirf** [reply]
Did you unprotect the flash successfully first?

**Reply by pauljsymonds** [reply]
I tried the unprotect command 1st

I have also tried on widows and the usb message there states malfunction device when connecting.

Device has 3 dashes on display

**Reply by marekhac** [reply]
Thanks @Lee-Adamson for your great instruction of flashing gotek under the Linux. I have one question that bothers me. When I try to unprotect the gotek before flashing with FlashFloppy, there is a strange thing. I execute the following command:

`dfu-util -a 0 -s 0x08000000:unprotect:force`

but it doesn't unlock the gotek. I need to execute this command once more. After that, the gotek is unblocked. 

Does anyone know why it is happening like this ? Why I need to run this command twice ?

**Reply by keirf** [reply]
It probably is deprotected on the first command despite presumably getting an error message. Deprotect causes mass erasure, and mass erase on these Artery chips is slow. So probably there is a timeout during the first command. And the second command ends up being a no-op and finishing immediately.

**Reply by marekhac** [reply]
> It probably is deprotected on the first command despite presumably getting an error message. Deprotect causes mass erasure, and mass erase on these Artery chips is slow. So probably there is a timeout during the first command. And the second command ends up being a no-op and finishing immediately.

Yeah, that make sense. Thanks Keir!

**Comment by niel** [comment]
Updating above instructions for my AT32F435 based Gotek.

Step 8 appears to be unnecessary now.
Step 10 My version comes with the I2C display which displays:
FlashFloppy+
version free_mem

Where version and free_mem are the actual values from the firmware/sram.

Many thanks for saving me a lot of time figuring this out ;-)

**Comment by shred** [comment]
Maybe it helps someone else: First attempt failed with an error `ERASE_PAGE not correctly executed` in step 9. I then just repeated step 8, and it worked after that.

**Reply by lizardb0y** [reply]
Thanks! I had the same issue. Step 8 returned the error:
```
DFU state(5) = dfuDNLOAD-IDLE, status(0) = No error condition is present
dfu-util: Wrong state after command "READ_UNPROTECT" download
```
Step 9 gave the error `ERASE_PAGE not correctly executed` several times. I repeated step 8 which returned the same error as before, but step 9 worked fine after that.

**Reply by Absolutely-Free** [reply]
I had the exact same issue, and it was resolved the same way. Thank you for sharing!

**Reply by gentleben** [reply]
Same here! For me it went step8=error, step9=error , step8=error, step9=error, step8=error, step9=success.

Thanks!

**Comment by janvrany** [comment]
Maybe it helps someone else: 

I tried above steps to flash friend's two STM32 based Goteks and failed (I succeeded using [serial](https://github.com/keirf/flashfloppy/wiki/Firmware-Programming#stm32-method-2-serial-programming) though!). Details below:

When plugged Gotek to USB, it took few seconds until it appeared on USB bus, this is what I've got in `dmesg`:
```
[43151.694997] usb 2-1-port5: attempt power cycle
[43152.298845] usb 2-1.5: new full-speed USB device number 29 using ehci-pci
[43152.714785] usb 2-1.5: device not accepting address 29, error -32
[43152.798780] usb 2-1.5: new full-speed USB device number 30 using ehci-pci
[43153.214794] usb 2-1.5: device not accepting address 30, error -32
[43153.214945] usb 2-1-port5: unable to enumerate USB device
[43158.802855] usb 2-1.5: new full-speed USB device number 31 using ehci-pci
[43158.882858] usb 2-1.5: device descriptor read/64, error -32
[43159.070855] usb 2-1.5: device descriptor read/64, error -32
[43159.258863] usb 2-1.5: new full-speed USB device number 32 using ehci-pci
[43159.338829] usb 2-1.5: device descriptor read/64, error -32
[43159.556982] usb 2-1.5: New USB device found, idVendor=0483, idProduct=df11, bcdDevice=22.00
[43159.556999] usb 2-1.5: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[43159.557004] usb 2-1.5: Product: STM32 0x418 DFU Bootloade
[43159.557007] usb 2-1.5: Manufacturer: STMicroelectronics
[43159.557010] usb 2-1.5: SerialNumber: STM32
```

Output from step 7:

```
<redacted>/flashfloppy-3.41$ sudo dfu-util --list
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Found Runtime: [0a5c:21e6] ver=0112, devnum=3, cfg=1, intf=3, path="1-1.4", alt=0, name="UNKNOWN", serial="20689D2ABFCD"
Found DFU: [0483:df11] ver=2200, devnum=32, cfg=1, intf=0, path="2-1.5", alt=1, name="@Option Bytes  /0x1FFFF800/01*016 e", serial="STM32"
Found DFU: [0483:df11] ver=2200, devnum=32, cfg=1, intf=0, path="2-1.5", alt=0, name="@Internal Flash  /0x08000000/128*002Kg", serial="STM32"
```

Output from step 8:

```
<redacted>/flashfloppy-3.41$ sudo dfu-util -D dfu/flashfloppy-at415-st105-3.41.dfu -a 0 -s 0x08000000:unprotect:force
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Match vendor ID from file: 0483
Match product ID from file: df11
Multiple alternate interfaces for DfuSe file
Opening DFU capable USB device...
Device ID 0483:df11
Device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Interface #0 ...
Determining device status...
DFU state(10) = dfuERROR, status(10) = Device's firmware is corrupt. It cannot return to run-time (non-DFU) operations
Clearing status
Determining device status...
DFU state(2) = dfuIDLE, status(0) = No error condition is present
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
Device disconnects, erases flash and resets now
```

Again, it took few seconds for USB device to reappear - but I guess that's normal. 

Step 9. fails with following:

```
<redacted>/flashfloppy-3.41$ sudo dfu-util -D dfu/flashfloppy-at415-st105-3.41.dfu -a 0
dfu-util 0.11

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2021 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Match vendor ID from file: 0483
Match product ID from file: df11
Multiple alternate interfaces for DfuSe file
Opening DFU capable USB device...
Device ID 0483:df11
Device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Interface #0 ...
Determining device status...
DFU state(10) = dfuERROR, status(10) = Device's firmware is corrupt. It cannot return to run-time (non-DFU) operations
Clearing status
Determining device status...
DFU state(2) = dfuIDLE, status(0) = No error condition is present
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
File contains 1 DFU images
Parsing DFU image 1
Target name: ST...
Image for alternate setting 0, (2 elements, total size = 116436)
Setting Alternate Interface #0 ...
Parsing element 1, address = 0x08000000, size = 30532
Erase   	[                         ]   0%            0 bytesdfu-util: dfuse_download: libusb_control_transfer returned -7 (LIBUSB_ERROR_TIMEOUT)
dfu-util: Error during special command "ERASE_PAGE" download
```
